lib/onigumo/parser.ex

```diff
@@ -0,0 +1,26 @@
+defmodule Onigumo.Parser do
+  @moduledoc """
+  Web scraper
+  """
+  @behaviour Onigumo.Component
+
+  @impl Onigumo.Component
+  def main(root_path) do
+    root_path
+    |> list_downloaded()
+    |> Enum.map(&IO.puts(&1))
+
+    :ok
+  end
+
+  defp list_downloaded(path) do
+    path
+    |> File.ls!()
```

Comment:
Consider, if the variant with exclamation mark is necessary to use. Maybe the elixir way of handling error will be better approach with state `{:ok} or {:err, msg}` will be better.

Reply:
I agree that we should avoid the bang functions. This is however a native function which we always expect to succeed. It is not our userspace function where we’d use the bang instead of a conditional.

We have #227 to test whether the path is valid. Now, we don’t do that even in the already existing Downloader: see the [`File.write!`](https://github.com/Glutexo/onigumo/blob/0d890b963ba18cca8d819bff58bd691adf27b458/lib/onigumo/downloader.ex#L43C14-L43C15) there.

With the path validation in place, any conditional here would imply that our validation may not work. I’d stick with the bang version then – even if it fails, it’d at least provide a useful error.